Add ADR to describe data validation (#885)
nordeck/matrix-widget-toolkit: docs/adrs/adr007-events-validation.md

> # ADR007: Events validation

Reviewer: ```suggestion
# ADR007: Data validation
```

I would say we should change it to data validation. In the description it is already mentioned, that the data source can be non-Matrix (thinking of the import for example). „events“ are very Matrix specific.

Reviewer 2: https://github.com/matrix-org/matrix.org/pull/2608

Reviewer 2: Please use title case
```suggestion
# ADR007: Event Validation
```
or with weeman's suggestion
```suggestion
# ADR007: Data Validation
```

Status: accepted

<!-- These documents have names that are short noun phrases. For example, "ADR001: Deployment on Ruby on Rails 3.0.10" or "ADR009: LDAP for Multitenant Integration" -->

## Context

Widgets are using Matrix events that are defined in Matrix specification and as well custom events that are
created and used by other custom matrix clients or widgets.

Matrix specification states that events should be considered as untrusted data. This applies both to Matrix and
custom events.

[MSC 2801][msc2801] provides reasons why events may not match the specification and the conclusion that
clients must treat all event data as untrusted.

Matrix events are described by matrix specification with [event schemas][eventSchemas]. This should be used as
a reference to implement validation logic of the events used.

Widgets may also load data from other data sources: localstorage, RTC channel, HTTP APIs. In general,
we could not consider these events and data to be of needed schemas all the time and therefore must validate.

<!--

> This section describes the forces at play, including technological, political, social, and project local. These forces are probably in tension, and should be called out as such. The language in this section is value-neutral. It is simply describing facts. -->

Author: Moved, thanks for finding

## Decision

<!-- This section describes our response to these forces. It is stated in full sentences, with active voice. "We will ..." -->

> We validate all events and data that are coming into widgets from different sources:

Reviewer 2: ```suggestion
We validate all events and data that is received by widgets from different sources:
```
small tweak?

For Matrix events (both from Matrix specification and custom widget events):

- `type` has expected event type, example: `m.room.name`, `net.nordeck.whiteboard`
- `content` to be an `object` and to be defined

> - `content` is validated to match the schema

Author: Widget is hosted by client. For example for element web some validations happen before event is sent to the widget: https://github.com/element-hq/element-web/blob/e14a3b64c386e568b14f1fbce7d3c9f5deb60677/src/stores/widgets/StopGapWidget.ts#L543-L578

Generally this client should validate these basic matrix fields. Synapse should also do some work here for these fields, but good point.

> - `content` is validated to match the schema

Reviewer 2: I think for this ADR, we need to
- either explain in a well-founded way why we don't need to validate any part that we don't validate.
  - E.g. in this case, why we can rely on Element to do the verification, by pointing to a part of the widget spec that requires it.
  - Or maybe because it makes no difference, e.g. if a field is never read? But I'm not sure about this option.
- or say we validate it anyway, to be on the safe side

> - `content` is validated to match the schema

Author: Valid point, this part of ADR is updated

For events and data from other data sources: we apply validation of data to match the schema.

Schema based validation includes:

- required fields are listed
- optional and required fields values have the expected data types
- fields primitive values match expected grammar: dates in ISO 8601, MXC URIs, whitelist values

> - fields object data is also validated according to schema

Author: Applied with changed to `field's`, a typo I guess.

We use several ways to validate incoming data.

### Plain code

> Validation logic could be written in plain code when it is not complicated or we don't want to use some library as external dependency.

Reviewer 2: > It is stated in full sentences, with active voice. "We will ..."

```suggestion
We write validation logic in plain code when it is not complicated or we don't want to use a library, introducing another external dependency.
```

> Example that validates that event has `m.relates_to` in the content:

Reviewer 2: ```suggestion
Example for validating that an event has an [`m.relates_to` object](https://spec.matrix.org/v1.13/client-server-api/#forming-relationships-between-events) in the content:
```

We could additionally permalink to the real place in the code for this. Is this a good idea?
`https://github.com/nordeck/matrix-widget-toolkit/blob/20d3c085c65ed063577b4b318aefb4b1610e3d8b/packages/api/src/api/extras/relatesTo.ts#L91-L121`

Author: yes, good Idea to provide more context for that. Links are added to two examples in adr.

<!--prettier-ignore-->
   ```ts
    // ....

    /**
     * Validates that `event` has a valid structure for a
     * {@link EventWithRelatesTo}.
     * @param event - The event to validate.
     * @returns True, if the event is valid.
     */
    export function isValidEventWithRelatesTo(
      event: RoomEvent,
    ): event is EventWithRelatesTo<string> {
      if (!event.content || typeof event.content !== 'object') {
        return false;
      }
    
      const relatedEvent = event as EventWithRelatesTo<string>;
    
      if (
        !relatedEvent.content['m.relates_to'] ||
        typeof relatedEvent.content['m.relates_to'] !== 'object'
      ) {
        return false;
      }
    
      if (
        typeof relatedEvent.content['m.relates_to'].rel_type !== 'string' ||
        typeof relatedEvent.content['m.relates_to'].event_id !== 'string'
      ) {
        return false;
      }
    
      return true;
    }
   ```

### Joi

> It is preferred to use [Joi][joi] when we have to validate data with more complex structure. Joi provides powerful

Reviewer 2: > It is stated in full sentences, with active voice. "We will ..."

```suggestion
We prefer to use [Joi][joi] when we need to validate data with more complex structure. Joi provides powerful
```

validation features.

Example that validates the content of `m.reaction.event`:

<!--prettier-ignore-->
   ```ts
    export type ReactionEvent = {
      'm.relates_to': {
        rel_type: string;
        event_id: string;
        key: string;
      };
    };
    
    const reactionEventSchema = Joi.object<ReactionEvent, true>({
      'm.relates_to': Joi.object({
        rel_type: Joi.string().required(),
        event_id: Joi.string().required(),
        key: Joi.string().required(),
      })
        .unknown()
        .required(),
    }).unknown();

    reactionEventSchema.validate(eventContent);
   ```

## Consequences

<!-- This section describes the resulting context, after applying the decision. All consequences should be listed here, not just the "positive" ones. A particular decision may have positive, negative, and neutral consequences, but all of them affect the team and project in the future. -->

We apply validation to all incoming Matrix events according to schemas and filter out invalid events. We also

> validate all other data that come from non Matrix data sources.

Reviewer 2: ```suggestion
validate all other data that comes from non-Matrix data sources.
```

> Matrix Widget Toolkit `api` package providers [validators][toolkitApiExtras]
> for some Matrix events to be used in widgets implementations.

Reviewer 2: ```suggestion
The Matrix Widget Toolkit `api` package provides [validators][toolkitApiExtras]
for some Matrix events to be used in widget implementations.
```

> Matrix Widget Toolkit `example-widget-mui` provides several [examples][toolkitExampleWidgetEvents] how to implementation validators using Joi.

Reviewer 2: ```suggestion
The Matrix Widget Toolkit `example-widget-mui` provides several [examples][toolkitExampleWidgetEvents] how to implement validators using Joi.
```
